python/tests/test_conversions.py

```diff
@@ -0,0 +1,422 @@
+        - 00010000 -> 16, 00100111 -> 39, ... , 00000000 -> 0
+    Timestamp:
+        - Stored as microseconds from 1970-01-01 00:00:00.000000 in an 8-byte little-endian long
+        - 400000L is 0...110|00011010|10000000 in binary
+        - 10000000 -> 128 (-128), 00011010 -> 26, 00000110 -> 6, ... , 00000000 -> 0
+    String:
+        - Stored as UTF-8 bytes (without length)
+        - 'A' -> 65, 'B' -> 66, 'C' -> 67
+    UUID:
+        - Stored as 16-byte big-endian values
+        - f79c3e09-677c-4bbd-a479-3f349cb785e7 is encoded as F7 9C 3E 09 67 7C 4B BD A4 79 3F 34 9C B7 85 E7
+        - 0xF7 -> 11110111 -> 247 (-9), 0x9C -> 10011100 -> 156 (-100), 0x3E -> 00111110 -> 62,
+        - 0x09 -> 00001001 -> 9, 0x67 -> 01100111 -> 103, 0x7C -> 01111100 -> 124,
+        - 0x4B -> 01001011 -> 75, 0xBD -> 10111101 -> 189 (-67), 0xA4 -> 10100100 -> 164 (-92),
+        - 0x79 -> 01111001 -> 121, 0x3F -> 00111111 -> 63, 0x34 -> 00110100 -> 52,
+        - 0x9C -> 10011100 -> 156 (-100), 0xB7 -> 10110111 -> 183 (-73), 0x85 -> 10000101 -> 133 (-123),
+        - 0xE7 -> 11100111 -> 231 (-25)
+    Fixed:
+        - Stored directly
+        - 'a' -> 97, 'b' -> 98
+    Binary:
+        - Stored directly
+        - 'Z' -> 90
+    Decimal:
+        - Stored as unscaled values in the form of two's-complement big-endian binary using the minimum number of bytes for the values
+        - 345 is 0...1|01011001 in binary
+        - 00000001 -> 1, 01011001 -> 89
+    Float:
+        - Stored as 4 bytes in little-endian order
+        - floating point numbers are represented as sign * 2ˆexponent * mantissa
+        - -4.5F is -1 * 2ˆ2 * 1.125 and encoded as 11000000|10010000|0...0 in binary
+        - 00000000 -> 0, 00000000 -> 0, 10010000 -> 144 (-112), 11000000 -> 192 (-64),
+"""
+import uuid
+from decimal import Decimal
+
+import pytest
+
+from iceberg import conversions
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+
+@pytest.mark.parametrize(
+    "primitive_type, partition_value_as_str, expected_result",
+    [
+        (BooleanType(), "true", True),
+        (BooleanType(), "false", False),
+        (BooleanType(), "TRUE", True),
+        (BooleanType(), "FALSE", False),
+        (BooleanType(), None, False),
+        (IntegerType(), "1", 1),
+        (IntegerType(), "9999", 9999),
+        (LongType(), "123456789", 123456789),
+        (FloatType(), "1.1", 1.1),
+        (DoubleType(), "99999.9", 99999.9),
+        (DecimalType(5, 2), "123.45", Decimal("123.45")),
+        (StringType(), "foo", "foo"),
+        (UUIDType(), "f79c3e09-677c-4bbd-a479-3f349cb785e7", uuid.UUID("f79c3e09-677c-4bbd-a479-3f349cb785e7")),
+        (FixedType(3), "foo", bytearray(b"foo")),
+        (BinaryType(), "foo", b"foo"),
+        (None, None, None),
```

Comment:
You may want a separate test that validates for all primitive types, None produces None.

Reply:
Add a [test_none_partition_values](https://github.com/apache/iceberg/pull/4263/files#diff-704ab9ed1661567f3b0eb10979903d6a9eb69efa2a76d099ed517ca19e5889beR143) test and it highlighted that a few weren't returning None for None so I updated those functions as well.